cudamapper/src/index_cache.cu

```diff
@@ -49,253 +126,281 @@ IndexCacheHost::IndexCacheHost(const bool same_query_and_target,
+                                                  kmer_size_,
+                                                  window_size_,
+                                                  hash_representations_,
+                                                  filtering_parameter_,
+                                                  cuda_stream_generation_,
+                                                  cuda_stream_copy_);
+
+            // wait for index to be generated on cuda_stream_generation_ before copying it on cuda_stream_copy_
+            // TODO: do this sync using an event
+            GW_CU_CHECK_ERR(cudaStreamSynchronize(cuda_stream_generation_));
+
+            if (host_copy_needed)
             {
-                // index already cached
-                index_on_host = existing_cache->second;
-                if (keep_on_device)
+                // if a D2H copy has been been started in the previous step wait for it to finish
+                // TODO: do this sync using an event
+                if (started_copy_from_previous_step)
                 {
-                    index_on_device = index_on_host->copy_index_to_device(allocator_, cuda_stream_);
+                    index_on_host_from_previous_step->finish_copying();
+                    // copying is done, these pointer are not needed anymore
+                    index_on_host_from_previous_step   = nullptr;
+                    index_on_device_from_previous_step = nullptr;
                 }
-            }
-            else
-            {
-                // create index
-                index_on_device = Index::create_index(allocator_,
-                                                      *parser,
-                                                      descriptor_of_index_to_cache,
-                                                      kmer_size_,
-                                                      window_size_,
-                                                      hash_representations_,
-                                                      filtering_parameter_,
-                                                      cuda_stream_);
-                // copy it to host memory
-                if (!skip_copy_to_host)
-                {
-                    index_on_host = IndexHostCopy::create_cache(*index_on_device,
+
+                index_on_host = IndexHostCopy::create_host_copy(*index_on_device,
                                                                 descriptor_of_index_to_cache.first_read(),
                                                                 kmer_size_,
                                                                 window_size_,
-                                                                cuda_stream_);
-                }
+                                                                cuda_stream_copy_);
+                started_copy  = true; // index is being copied to host memory, a sync will be needed
```

Comment:
Can't you just set `index_on_host_from_previous_step` and `index_on_device_from_previous_step` here?
Then you should be able to avoid the `started_copy` bool completely, by just checking `index_on_host_from_previous_step != nullptr`.
You may then also allow for a copy to take multiple iterations of this loop - until a new copy needs to be initiated. You should also be able to remove the whole logic between lines 246-266.
Optional: If you make the change, I would also consider to rename these variables to `index_on_host_copying_in_progress` or similar.

Reply:
Optional: How about `index_on_host_copy_in_flight`? :)

Reply:
I don't think this would work. Check on line 247 cannot be avoided as it can happen that in one iteration copy is started and then in the next iteration index is found in cache, so the sync on line 217 would be skipped.
As the check on like 247 cannot be avoided `index_on_host_from_previous_step` cannot be reset as the application would still need the value from the previous iteration.

I agree that this function is rather complicated. I'm thinking whether it is a bit overengineered for this use case as it assumes that every index can independently be found it his, the other cache or not be found at all. In reality indices are split into batches that are almost always the same. For example if there are 10 indices and 3 batches, with default batch size 4 there would be batches (0, 3), (4, 7) and (8, 9). This means that for every cache we could keep an explicit list of cached indices and only check if the new cache requests exactly the same indices.
This would simplify the code a lot (we either know all indices are already cached or all of them have to created and potentially copied to host), but would lose performance in more general case by regenerating and copying indices which already exist.

Reply:
After a call it turned out that this can be done.
done